lib/SILOptimizer/Mandatory/MandatoryInlining.cpp

```diff
@@ -42,146 +45,317 @@ static void diagnose(ASTContext &Context, SourceLoc loc, Diag<T...> diag,
+    if (v->getType().isTrivial(pai->getModule()))
+      continue;
+
+    bool hasOwnership = applySite->getFunction()->hasOwnership();
+
+    switch (convention) {
+    case ParameterConvention::Indirect_In:
+    case ParameterConvention::Indirect_In_Constant:
+    case ParameterConvention::Indirect_Inout:
+    case ParameterConvention::Indirect_InoutAliasable:
+    case ParameterConvention::Indirect_In_Guaranteed:
+      llvm_unreachable("Should be handled above");
+
+    case ParameterConvention::Direct_Guaranteed: {
+      // If we have a direct_guaranteed value, the value is being taken by the
+      // partial_apply at +1, but we are going to invoke the value at +0. So we
+      // need to copy/borrow the value before the pai and then
+      // end_borrow/destroy_value at the apply site.
+      SILValue copy = SILBuilderWithScope(pai).emitCopyValueOperation(loc, v);
+      SILValue argument = copy;
+      if (hasOwnership) {
+        argument = SILBuilderWithScope(pai).createBeginBorrow(loc, argument);
+      }
+
+      visitedBlocks.clear();
+      // If we need to insert compensating destroys, do so.
+      if (!valueHasLinearLifetime(copy, {applySite}, {}, visitedBlocks,
+                                  deadEndBlocks, errorBehavior,
+                                  &leakingBlocks)) {
+        while (!leakingBlocks.empty()) {
+          auto *leakingBlock = leakingBlocks.pop_back_val();
+          auto loc = RegularLocation::getAutoGeneratedLocation();
+          SILBuilderWithScope builder(leakingBlock->begin());
+          builder.emitDestroyValueOperation(loc, copy);
+        }
+      }
+
+      insertAfterApply(applySite, [&](SILBasicBlock::iterator iter) {
+        if (hasOwnership) {
+          SILBuilderWithScope(iter).createEndBorrow(loc, argument);
+        }
+        SILBuilderWithScope(iter).emitDestroyValueOperation(loc, copy);
+      });
+      v = argument;
+      break;
+    }
+
+    // TODO: Do we need to lifetime extend here?
+    case ParameterConvention::Direct_Unowned: {
+      v = SILBuilderWithScope(pai).emitCopyValueOperation(loc, v);
+
+      visitedBlocks.clear();
+      // If we need to insert compensating destroys, do so.
+      if (!valueHasLinearLifetime(v, {applySite}, {}, visitedBlocks,
+                                  deadEndBlocks, errorBehavior,
+                                  &leakingBlocks)) {
+        while (!leakingBlocks.empty()) {
+          auto *leakingBlock = leakingBlocks.pop_back_val();
+          auto loc = RegularLocation::getAutoGeneratedLocation();
+          SILBuilderWithScope builder(leakingBlock->begin());
+          builder.emitDestroyValueOperation(loc, v);
+        }
+      }
+
+      insertAfterApply(applySite, [&](SILBasicBlock::iterator iter) {
+        SILBuilderWithScope(iter).emitDestroyValueOperation(loc, v);
+      });
+      break;
+    }
+
+    // If we have an owned value, we need to insert a copy here for lifetime
+    // extension purposes.
```

Comment:
Could we improve the comment: If we have an owned value, we insert the copy for two purposes: first, to balance the consuming argument, this also has the effect to satisfy the second purpose of extending the lifetime.